### PR TITLE
Add Sync Session ID Synchronizer State

### LINKED ISSUE #895: Add Sync Session UUID

### PR DESCRIPTION
**Pull request by pacu**
Closes #895 

Add Sync Session ID to `SynchronizerState`.

A SyncSession is an attempt to sync the blockchain within the lifetime of a Synchronizer.
A Synchronizer can sync many times, when synced it will refresh every ~20 seconds +- random padding.
each sync attempt will have a different UUID even if it's from the same instance of SDKSynchronizer.

**How are SyncSessions are delimited? With `SyncStatus` changes.** 

changes from [`.unprepared`|`.error`|`.disconnected`|`.stopped`] to `.syncing` assign a new sessionID to the synchronizer.

Any other transitions won't. 

`areTwoStatusesDifferent ` was refactored to a helper function of `SyncStatus`

**How are IDs generated?**

ID generation is not mandated but delegated to a protocol `SyncSessionIDGenerator`.
Tests inject their own deterministic generator to avoid test flakiness.

Default implementation of SyncSessionIDGenerator is 

````Swift
struct UniqueSyncSessionIDGenerator {}

extension UniqueSyncSessionIDGenerator: SyncSessionIDGenerator {
    func nextID() -> UUID {
        UUID()
    }
}
````

**SyncSession Pseudo-Atomicity and thread safety** 

SyncSession is a type alias of a GenericActor holding a UUID

````
typealias SyncSession = GenericActor<UUID>

extension SyncSession {
    /// updates the current sync session to a new value with the given generator
    /// - Parameters generator: a `SyncSessionIDGenerator`
    /// - returns: the `UUID` of the newly updated value.
    @discardableResult
    func newSession(with generator: SyncSessionIDGenerator) async -> UUID {
        return await self.update(generator.nextID())
    }
}
````
This code review checklist is intended to serve as a starting point for the author and reviewer, although it may not be appropriate for all types of changes (e.g. fixing a spelling typo in documentation).  For more in-depth discussion of how we think about code review, please see [Code Review Guidelines](../blob/main/CODE_REVIEW_GUIDELINES.md).

# Author

- [x] Self-review: Did you review your own code in GitHub's web interface? Code often looks different when reviewing the diff in a browser, making it easier to spot potential bugs.
- [ ] Automated tests: Did you add appropriate automated tests for any code changes?
- [ ] Code coverage: Did you check the code coverage report for the automated tests?  While we are not looking for perfect coverage, the tool can point out potential cases that have been missed.
- [ ] Documentation: Did you update Docs as appropiate? (E.g [README.md](../blob/main/README.md), etc.)
- [ ] Run the app: Did you run the app and try the changes? 
- [ ] Did you provide Screenshots of what the App looks like before and after your changes as part of the description of this PR? (only applicable to UI Changes)
- [x] Rebase and squash: Did you pull in the latest changes from the main branch and squash your commits before assigning a reviewer? Having your code up to date and squashed will make it easier for others to review. Use best judgement when squashing commits, as some changes (such as refactoring) might be easier to review as a separate commit.

# Reviewer

- [ ] Checklist review: Did you go through the code with the [Code Review Guidelines](../blob/main/CODE_REVIEW_GUIDELINES.md) checklist?
- [ ] Ad hoc review: Did you perform an ad hoc review?  _In addition to a first pass using the code review guidelines, do a second pass using your best judgement and experience which may identify additional questions or comments. Research shows that code review is most effective when done in multiple passes, where reviewers look for different things through each pass._
- [ ] Automated tests: Did you review the automated tests?
- [ ] Manual tests: Did you review the manual tests?_You will find manual testing guidelines under our [manual testing section](../blob/mater/docs/testing/manual_testing)_
- [ ] How is Code Coverage affected by this PR? _We encourage you to compare coverage befor and after your changes and when possible, leave it in a better place. [Learn More...](../blob/master/docs/testing/local_coverage.md)_
- [ ] Documentation: Did you review Docs, [README.md](../blob/master/README.md), [LICENSE.md](../blob/master/LICENSE.md), and [Architecture.md](../blob/master/docs/Architecture.md) as appropriate?
- [ ] Run the app: Did you run the app and try the changes? While the CI server runs the app to look for build failures or crashes, humans running the app are more likely to notice unexpected log messages, UI inconsistencies, or bad output data.

### CODE REVIEW

File: Sources/ZcashLightClientKit/Synchronizer.swift [hunk 1]
```diff
@@ -64,15 +64,27 @@ public enum ConnectionState {
...
 }
-
+/// Reports the state of a synchronizer.
 public struct SynchronizerState: Equatable {
+    /// Unique Identifier for the current sync attempt
+    /// - Note: Although on it's lifetime a synchronizer will attempt to sync between random fractions of a minute (when idle),
+    /// each sync attempt will be considered a new sync session. This is to maintain a consistent UUID cadence
+    /// given how application lifecycle varies between OS Versions, platforms, etc.
+    /// SyncSessionIDs are provided to users
+    public var syncSessionID: UUID?
```
Comment (Chlup) [inline Sources/ZcashLightClientKit/Synchronizer.swift:74]: Why is this Optional? Maybe I don't see some context but we can generate UUID anytime without any Optional. So I think state can have ID anytime. Can't it? What are the cases when this is nil? Even the "placeholder" property `.zero` uses some ID.

File: Sources/ZcashLightClientKit/Synchronizer/SDKSynchronizer.swift [hunk 2]
```diff
@@ -583,6 +590,27 @@ public class SDKSynchronizer: Synchronizer {
...
+    /// Helper function to determine whether we are in front of a SyncSession change for a given syncStatus
+    /// transition we consider that every sync attempt is a new sync session and should have it's unique UUID reported.
+    static func isNewSyncSession(oldStatus: SyncStatus, newStatus: SyncStatus) -> Bool {
+        // if the state hasn't changed to a different syncStatus member
+        guard oldStatus.isDifferent(from: newStatus) else { return false }
+
+        switch (oldStatus, newStatus) {
+        case (.unprepared, .syncing):
+            return true
+        case (.unprepared, _):
```
Comment (Chlup) [inline Sources/ZcashLightClientKit/Synchronizer/SDKSynchronizer.swift:602]: This case can be handled in default. Is there any motivation to explicitly state it here? It works fine. I just seem little bit unusual.

File: Sources/ZcashLightClientKit/Synchronizer/SDKSynchronizer.swift [hunk 3]
```diff
@@ -592,12 +620,23 @@ public class SDKSynchronizer: Synchronizer {
         // When new snapshot is created balance is checked. And when balance is checked and data DB doesn't exist then rust initialise new database.
         // So it's necessary to not create new snapshot after status is switched to `unprepared` otherwise data DB exists after wipe
         if newStatus == .unprepared {
-            newState = SynchronizerState.zero
+            var nextState = SynchronizerState.zero
+
+            let nextSessionID = await self.syncSession.newSession(with: self.syncSessionIDGenerator)
+
+            nextState.syncSessionID = nextSessionID
```
Comment (Chlup) [inline Sources/ZcashLightClientKit/Synchronizer/SDKSynchronizer.swift:627]: In the description of the PR you state that session ID changes when status is switched from something to `syncing`. But here status switches to `unprepared` and new ID is generated. Now the SDK behave little bit different when it's created (status is unprepared and sync ID is zero) and when it's wiped (status is unprepared and sync ID is real). Maybe this is OK and what we want. It is just little discrepancy with description so I want to be sure that this is intention.

File: Sources/ZcashLightClientKit/Synchronizer/SDKSynchronizer.swift [hunk 4]
```diff
@@ -583,6 +590,27 @@ public class SDKSynchronizer: Synchronizer {
         )
     }
 
+    /// Helper function to determine whether we are in front of a SyncSession change for a given syncStatus
+    /// transition we consider that every sync attempt is a new sync session and should have it's unique UUID reported.
+    static func isNewSyncSession(oldStatus: SyncStatus, newStatus: SyncStatus) -> Bool {
```
Comment (Chlup) [inline Sources/ZcashLightClientKit/Synchronizer/SDKSynchronizer.swift:595]: Why is this function static?

File: Tests/DarksideTests/SynchronizerDarksideTests.swift [hunk 5]
```diff
@@ -136,7 +137,7 @@ class SychronizerDarksideTests: XCTestCase {
             error: self.handleError
         )
         
-        wait(for: [findManyTxExpectation], timeout: 10)
+        await fulfillment(of: [findManyTxExpectation], timeout: 10)
```
Comment (Chlup) [inline Tests/DarksideTests/SynchronizerDarksideTests.swift:140]: If `fulfillment` is used then Github action doesn't work. So we either have to deal with warnings or use `fulfillment` and deal with CI failures until github update their thing. I am just stating this so we are aware of the consequences.

**Review by Chlup — CHANGES_REQUESTED** [review]

File: Sources/ZcashLightClientKit/Synchronizer.swift (hunk 1 above)
Comment (pacu) [reply Sources/ZcashLightClientKit/Synchronizer.swift:74]: thanks for pointing this out! I forgot to remove the optional.

**Review by Chlup — APPROVED** [review]

**Pull request merged**